engine/new_runtime/src/memories/file_memory.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    crypto_utils::crypto_box::{BoxProvider, Key},
+    locked_memory::{
+        LockedConfiguration::{self, *},
+        LockedMemory,
+        MemoryError::{self, *},
+        ProtectedConfiguration, ProtectedMemory,
+    },
+    memories::buffer::Buffer,
+    types::ContiguousBytes,
+};
+use core::fmt::{self, Debug, Formatter};
+use rand_ascii::{distributions::Alphanumeric, thread_rng, Rng};
+use std::{
+    fs::{self, File},
+    io::prelude::*,
+};
+use zeroize::Zeroize;
+
+static ERR_SIZE_NONE: &'static str = "FileMemory: the size should not be None if allocated properly";
+
+const FILENAME_SIZE: usize = 16;
+const AD_SIZE: usize = 32;
+
+/// File memory
+pub struct FileMemory<P: BoxProvider> {
+    // Filename are random string of 16 characters
+    fname: String,
+    config: LockedConfiguration<P>,
+    // Nonce for encrypted memory
+    ad: [u8; AD_SIZE],
+}
+
+impl<P: BoxProvider> FileMemory<P> {
+    fn random_fname() -> String {
+        thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(FILENAME_SIZE)
+            .map(char::from)
+            .collect()
+    }
+
+    // TODO: add security
+    // - file permissions
+    // - noise in the file
+    fn read_file(&self) -> Result<Vec<u8>, std::io::Error> {
+        fs::read(&self.fname)
+    }
+
+    // TODO: add security
+    // - file permissions
+    // - noise in the file
+    fn write_to_file(payload: &[u8]) -> Result<String, std::io::Error> {
+        let fname: String = FileMemory::<P>::random_fname();
+        let mut file = File::create(&fname)?;
+        file.write_all(payload.as_bytes())?;
```

Comment:
This is an amazing Pull Request. My only comment is that you might like to consider writing to a temporary file and then mv'ing it to its final name and performing a sync. Back in the day we saw some corruption with snapshots, and this is the way it was solved:

https://github.com/iotaledger/stronghold.rs/blob/1dfd61b9c393675173bb2e24a6e5b04d2a4bc473/engine/src/snapshot/logic.rs#L190-L196

Reply:
Hey @nothingismagick, thanks for the nice words, but you are actually referring work from this [PR#316](https://github.com/iotaledger/stronghold.rs/pull/316). @vuongDang deserve the credit for implementing all the memory features :blush: 

I am only pulling in work from this branch to build on top of it.